decidim-core/app/services/decidim/onboarding_manager.rb

```diff
@@ -0,0 +1,229 @@
+# frozen_string_literal: true
+
+module Decidim
+  # This class is used to manage the onboarding funnel for users that
+  # intended to perform an action in the platform but it requires some
+  # verification steps to be completed.
+  #
+  # The flow is as follows:
+  #
+  # 1. A visitor (logged out) tries to perform an action that requires to be logged in
+  # 2. The intended action is stored in a cookie (onboarding key)
+  # 3. The visitor logs in and the onboarding cookie data is merged into the User#extended_data column.
+  # 4. After logging in, the user is redirected to the onboarding pending authorizations page
+  #    if there is any pending action stored in the User#extended_data column
+  # 5. The user completes the required authorizations
+  # 6. When the onboarding authorizations page detects that all the verification steps have been
+  #    completed, it will remove the pending action from the User#extended_data column and redirect
+  #    the user to the intended action.
+  #
+  class OnboardingManager
+    attr_reader :user
+
+    def initialize(user)
+      @user = user
+    end
+
+    # Checks if the onboarding data has an action and a model.
+    #
+    # Returns a boolean
+    def valid?
+      return if action.blank?
+
+      permissions_holder.present?
+    end
+
+    # Returns the action to be performed.
+    #
+    # Returns a string
+    def action
+      onboarding_action
+    end
+
+    # Returns the translation of the action to be performed if found, action
+    # elsewhere
+    #
+    # Returns a string
+    def action_text
+      @action_text ||= if component && I18n.exists?("#{component.manifest.name}.actions.#{action}", scope: "decidim.components")
+                         I18n.t("#{component.manifest.name}.actions.#{action}", scope: "decidim.components")
+                       elsif permissions_holder.respond_to?(:resource_manifest) &&
+                             I18n.exists?("#{permissions_holder.resource_manifest.name}.actions.#{action}", scope: "decidim.resources")
+                         I18n.t("#{permissions_holder.resource_manifest.name}.actions.#{action}", scope: "decidim.resources")
+                       else
+                         action
+                       end
+    end
+
+    # Checks if there is any pending action for the user.
+    #
+    # Returns a boolean
+    def pending_action?
+      onboarding_data.present?
+    end
+
+    # Checks if the user has completed all the required authorizations.
+    #
+    # active_authorization_methods - A list of the active authorization methods for the user.
+    #
+    # Returns a boolean
+    def finished_verifications?(authorization_methods = active_authorization_methods)
+      (authorization_handlers - authorization_methods).empty?
+    end
+
+    # Removes the pending action from the user's extended_data.
+    #
+    # Returns nothing
+    def remove_pending_action!
```

Comment:
This is duplicated in the concern `OnboardingActionMethods`. Given this service is used in a controller too, won't it be possible to avoid the duplication and use the concern method `clear_onboarding_data!(user)` to delete the pending action?

Reply:
You are right!